### PR TITLE
Allow Users to use the bot api

### PR DESCRIPTION
**Pull request by code1mountain**
Hey,

I received all your feedback on my last implementation of user mode and worked hard to deliver. The significant changes are:

1. The user token is now created by the server with 256bit of random data and has the format `integer:base64url_data`. This makes sure that no insecure tokens are used and it passes the client side token validations.
2. The 2fa password is now passed in a separate authentication step, so it does not need to be stored on the server.
3. Added a user database to ensure that users and bots can only use the correct endpoint.

Read the Documentation in the Readme for the new authorization process.

Please give feedback on this pull request. I plan to add user only methods after this Pull Request is accepted.

### CODE REVIEW

**Comment by MarcoBuster** [comment]
Thank you very much for your awesome work! Can I close the other PR? (#18)

**Comment by code1mountain** [comment]
@MarcoBuster Yes, this new PR replaces the old one

File: telegram-bot-api/Client.cpp
Comment (giuseppeM99) [inline telegram-bot-api/Client.cpp:198]: Shouldn't this default to false?
Since the user mode is a new addition we should default to the previous mode (that is assume is a bot)

**Comment by MarcoBuster** [comment]
Also, for the next commits, I suggest you (@code1mountain) to set up the committer name and email correctly on your git client because now it's not right

**Comment by iiiiii1wepfj** [comment]
add methods like in pyrogram

**Comment by giuseppeM99** [comment]
> add methods like in pyrogram

new methods will be implemented in the future, this PR is for the user mode, if you wish to suggest some new features open a new issue

File: telegram-bot-api/Client.cpp
Comment (code1mountain) [reply telegram-bot-api/Client.cpp:198]: I have to constructors for Client. This constructor is only called in the userlogin method and passes a phone number, so it is automatically a user. The other constructor is used when logging in as a bot.

Comment (giuseppeM99) [reply telegram-bot-api/Client.cpp:198]: i'd prefer to say explicitally if it is an user and assume that is a bot otherwise to preserve the original functionalities (as it may break the upstream)
awaiting for @andrew-ld to review this

Comment (code1mountain) [reply telegram-bot-api/Client.cpp:198]: The other Constructor explicitely says if it is an user and this Constructor can not be called for bots because of the phone number argument. But I can create a parameter that always passes true for is_user.

File: README.md
Comment (MarcoBuster) [inline README.md:131]: Read the previous comment.

**Review by MarcoBuster — CHANGES_REQUESTED** [review]
I only reviewed the docs and the code formatting; an in-depth review about the functionality will come later: this PR can't be merged without extensive tests and other reviews by core members. Many of the comments are personal opinions, maybe you or some other contributor have different ones; discuss in the comments if needed, I'm open to feedback. We're glad for the work and effort you've put into this, thank you!

File: README.md
Comment (giuseppeM99) [reply README.md:18]: I'd use "user mode"

File: README.md
Comment (code1mountain) [reply README.md:153]: This is basically just a typo, it is the same method as in the normal bot api, did not touch it.

File: telegram-bot-api/telegram-bot-api.cpp
Comment (giuseppeM99) [inline telegram-bot-api/telegram-bot-api.cpp:178]: Do not remove this line

**Review by giuseppeM99 — CHANGES_REQUESTED** [review]

File: telegram-bot-api/telegram-bot-api.cpp
Comment (code1mountain) [reply telegram-bot-api/telegram-bot-api.cpp:178]: I just reverted it the way it was before my pull request.

File: telegram-bot-api/Client.cpp
Comment (andrew-ld) [inline telegram-bot-api/Client.cpp:7708]: last_name should never be empty, rather put it as null, because telegrams block the signup under certain circumstances if you enter an empty string as last_name

**Review by andrew-ld — CHANGES_REQUESTED** [review]

**Review by Cavallium — APPROVED** [review]

File: telegram-bot-api/Client.cpp
Comment (Cavallium) [reply telegram-bot-api/Client.cpp]: false, it must not be null, otherwise it will ban the user!

Comment (Cavallium) [reply telegram-bot-api/Client.cpp]: An empty last name must be an empty string

**Review by Cavallium — CHANGES_REQUESTED** [review]
Last name must not be null

**Comment by andrew-ld** [comment]
> Last name must not be null

@code1mountain I confirm what cavallium say, we have discussed internally, and indeed the problem is when it is null and not when it is empty.

File: telegram-bot-api/Client.h
Comment (MarcoBuster) [reply telegram-bot-api/Client.h:511]: Oh, then it's ok

File: telegram-bot-api/telegram-bot-api.cpp
Comment (giuseppeM99) [reply telegram-bot-api/telegram-bot-api.cpp:178]: This option is still not present, please readd the relative optoin

File: telegram-bot-api/telegram-bot-api.cpp
Comment (giuseppeM99) [inline telegram-bot-api/telegram-bot-api.cpp:177]: This is a duplicate, remove

**Review by giuseppeM99 — CHANGES_REQUESTED** [review]

File: README.md
Comment (MarcoBuster) [reply README.md:18]: `--allow-users` command line option I think it's ok, but I personally prefer "user mode" / "user login" for documentation

File: telegram-bot-api/telegram-bot-api.cpp
Comment (code1mountain) [reply telegram-bot-api/telegram-bot-api.cpp:178]: Done, sorry

File: README.md
Comment (code1mountain) [reply README.md]: Okay, renamed

File: README.md
Comment (code1mountain) [reply README.md]: I renamed it to User Mode in the docs

**Review by Cavallium — APPROVED** [review]

**Review by Cavallium — APPROVED** [review]

File: telegram-bot-api/Client.cpp
Comment (giuseppeM99) [reply telegram-bot-api/Client.cpp:198]: @code1mountain change `is_user_(true)` to `is_user_(false)` and call the explicit costructor when the user mode is needed, this way we can preserve backward compatibility

**Review by giuseppeM99 — CHANGES_REQUESTED** [review]

**Review by andrew-ld — APPROVED** [review]

File: telegram-bot-api/Client.cpp
Comment (code1mountain) [reply telegram-bot-api/Client.cpp]: I think there is a misunderstanding: The two Constructors are for different tasks.

1. The first Constructor is for waking up already authorized Clients or creating new Bot Clients. This is the default Constructor that has a new argument `is_user`

2. (This one) The second Constructor is only used when creating a new User Client. Therefore it can be automatically assumed that the Client is a User.

**Review by Cavallium — APPROVED** [review]

**Review by giuseppeM99 — APPROVED** [review]

**Review by MarcoBuster — CHANGES_REQUESTED** [review]
After further testing, I was not able to login properly using an instance built on the latest commit of your branch; @giuseppeM99 reported issues with webhooks when migrating from an older version. We discussed the issues on the Telegram group and in private, but at this state we're not able to merge until we figure out those bugs.

**Comment by giuseppeM99** [comment]
Marking this PR as a work in progress

**Review by Cavallium — APPROVED** [review]

**Comment by JosXa** [comment]
Is there any consideration towards "how easy we're making it for scammers and spammers to add people to groups" and the whole "easily-accessible userbots commoditize power" story, or do we simply trust on the fact that phone numbers aren't exactly free?

**Comment by code1mountain** [comment]
> Is there any consideration towards "how easy we're making it for scammers and spammers to add people to groups" and the whole "easily-accessible userbots commoditize power" story, or do we simply trust on the fact that phone numbers aren't exactly free?

There are ready to use userbots for scammers to add users to groups. I think there already exist much easier ways than setting up this api and running a bot on it.

**Comment by MarcoBuster** [comment]
> Is there any consideration towards "how easy we're making it for scammers and spammers to add people to groups" and the whole "easily-accessible userbots commoditize power" story, or do we simply trust on the fact that phone numbers aren't exactly free?

As @code1mountain said, creating malicious userbots today is already very easy and we're not censoring features and code because it may be used to do so. Even bots can be malicious, should we stop the entire project then? Makes no sense.

**Review by giuseppeM99 — APPROVED** [review]

**Comment by code1mountain** [comment]
All known issues are fixed now. I think this is ready to merge after some more testing.

**Comment by giuseppeM99** [comment]
I've been testing it and looks good for me
We just need to wait for @MarcoBuster approval

**Review by MarcoBuster — APPROVED** [review]
LGTM :)

**Comment by MarcoBuster** [comment]
Green light. Let's merge and hope for the best.
let's gooo

Again, thank you for this awesome pull request and your time. Hope to see more from you in the future.

**Comment by JosXa** [comment]
Congratulations! 🔥 Well done, good job and thank you everyone, this is (going to be) incredibly useful and impactful!

**Pull request merged**
